bin/syclcc-clang

```diff
@@ -448,135 +494,239 @@ def run_or_print(command, print_only):
-    ]
+  def get_linker_flags(self):
+    return self._linker_args
 
-  def _get_rocm_clang_include_path(self, config):
-    return config.clang_include_path
+class omp_invocation:
+  def __init__(self, config):
+    if not "omp" in config.targets:
+      raise RuntimeError("OpenMP not among targets")
+    if len(config.targets["omp"]) != 0:
+      raise RuntimeError("OpenMP backend does not support specifiying target architecture")
 
-  def run(self):
-    command = [self._clang] + self._common_compiler_args
+    self._cxx = config.pure_cpu_compiler
 
-    if self._contains_compilation_stage:
-      command += self._compiler_args
+  def get_compiler_preference(self):
+    return (self._cxx, 1)
 
-    if self._contains_linking_stage:
-      command += self._linker_args
-    
-    command += self._args
-    return run_or_print(command, self._is_dry_run)
-    
-    
+  def get_cxx_flags(self):
 
-class pure_cpu_compiler:
-  def __init__(self, config):
-    self._args = config.forwarded_compiler_arguments
-    self._contains_linking_stage = config.contains_linking_stage()
-    self._is_dry_run = config.is_dryrun
+    flags = [
+        "-D__HIPSYCL_ENABLE_OMPHOST_TARGET__",
+      ]
 
-    try:
-      # First, see if any value for the pure cpu compiler
-      # is specified...
-      self._compiler = config.pure_cpu_compiler
-    except:
-      # if not, fall back to clang
-      self._compiler = config.clang_path
+    if sys.platform == "darwin":
+        # y'all still need libomp installed
+        flags += ["-Xclang"]
+    flags += ["-fopenmp"]
       
-    hipsycl_library_path = os.path.join(config.hipsycl_installation_path,"lib/")
+    return flags
 
-    self._linker_args = []
-    self._compiler_args = [
-      "-D__HIPSYCL__",
-      "-D__HIPSYCL_ENABLE_OMPHOST_TARGET__"
+  def get_linker_flags(self):
+    linker_args = [
+      "-lboost_context",
+      "-lboost_fiber"
     ]
+
     if sys.platform == "darwin":
         # y'all still need libomp installed
-        self._compiler_args += ["-Xclang"]
-    self._compiler_args += ["-fopenmp"]
-    self._compiler_args += config.common_compiler_args
+        linker_args += ["-Xclang"]
+    linker_args += ["-fopenmp"]
+
+    return linker_args
+
+# This is a workaround to have access to a backend
+# that can execute host tasks when compiling for GPUs.
+# It should be removed once we have non-OpenMP host backends
+# (e.g. TBB)
+class omp_sequential_invocation:
+  def __init__(self, config):
+    self._cxx = config.pure_cpu_compiler
+
+  def get_compiler_preference(self):
+    return (self._cxx, 1)
 
-    self._linker_args += [
-      "-L"+hipsycl_library_path,
-      "-lhipSYCL-rt",
+  def get_cxx_flags(self):
+
+    flags = [
+        "-D__HIPSYCL_ENABLE_OMPHOST_TARGET__",
+      ]
+      
+    return flags
+
+  def get_linker_flags(self):
+    linker_args = [
       "-lboost_context",
-      "-lboost_fiber"
+      "-lboost_fiber",
+      "-lomp"
```

Comment:
While temporary, this presumes the LLVM implementation of OpenMP, which is not by default included in LLVM/Clang installs, and will likely conflict if the user selects something other than Clang as the host compiler. As a primarily ROCm user, I'm not expecting a SYCL implementation to make any particular demands of the OpenMP implementation just to run GPU code, though GCC 8 suffices for the C++17 OpenMP (4.5 at least) requirements.

I'm working on a patch to infer the OpenMP flags that correspond with whatever `cpu-compiler` is set to. Though it could also be implemented as an explicit parameter if there is a strong need to have it differ from the one corresponding to the host compiler.

Reply:
> While temporary, this presumes the LLVM implementation of OpenMP, which is not by default included in some LLVM/Clang installs, and will likely conflict if the user selects something other than Clang as the host compiler

No - this is the omp sequential backend which is not actually used when compiling for CPU. It is used when compiling for *GPU* backends so that we are guaranteed that we always have a CPU backend available that can for example run SYCL 2020 host tasks. Effectively it is the OpenMP backend without `-fopenmp`.

The reason we don't use the regular OpenMP backend here is that there is a bug in clang CUDA that prevents `clang -x cuda -fopenmp` to work.

Because it is only used when also compiling for GPU, we are always guaranteed that we are using clang.

>  I'm not expecting a SYCL implementation to make any particular demands of the OpenMP implementation just to run GPU code

hipSYCL requires OpenMP always and this is enforced by cmake. As mentioned previously even when you just want to run on GPU hipSYCL still needs a CPU backend, for example for SYCL 2020 host tasks. Or for the case when a user requests data copies between two host accessors. This is also why you cannot build hipSYCL without the CPU backend enabled.

My plan for the future is to add something like a TBB backend that we can use as default CPU backend instead of openmp sequential, which is basically a hack.

Reply:
Hmmm

> > While temporary, this presumes the LLVM implementation of OpenMP, which is not by default included in some LLVM/Clang installs, and will likely conflict if the user selects something other than Clang as the host compiler
> 
> No - this is the omp sequential backend which is not actually used when compiling for CPU. It is used when compiling for _GPU_ backends so that we are guaranteed that we always have a CPU backend available that can for example run SYCL 2020 host tasks. Effectively it is the OpenMP backend without `-fopenmp`.

So I can see that *if* OpenMP is selected as one of the targets, it uses the full `-fopenmp` rather than getting the link flag from `omp_sequential_invocation`. But even then it does not use the OpenMP of the user-requested host compiler, at least in my testing. `fopenmp` always presumes `libomp` unless I modify it to `-fopenmp=libgomp`, which the HIP backend then accepts. Thus the idea to pin the OpenMP flags to match the host compiler.

> 
> The reason we don't use the regular OpenMP backend here is that there is a bug in clang CUDA that prevents `clang -x cuda -fopenmp` to work.

This makes perfect sense until that upstream clang bug is resolved. Until then LLVM OpenMP probably needs to be added as an explicit dependency for the CUDA backend.
> 
> Because it is only used when also compiling for GPU, we are always guaranteed that we are using clang.
> 
> > I'm not expecting a SYCL implementation to make any particular demands of the OpenMP implementation just to run GPU code
> 
> hipSYCL requires OpenMP always and this is enforced by cmake. As mentioned previously even when you just want to run on GPU hipSYCL still needs a CPU backend, for example for SYCL 2020 host tasks. Or for the case when a user requests data copies between two host accessors. This is also why you cannot build hipSYCL without the CPU backend enabled.

Yeah the need for a host implementation is explicit in the SPEC, so the need for OpenMP in lieu of TBB, pthreads, etc makes sense. Cmake is finding my GCC OpenMP so it's not failing, but not identifying the lack of LLVM OpenMP. Paths corresponding to the GCC implementation don't get forwarded to clang during the GPU passes, so it chokes looking for headers.

> 
> My plan for the future is to add something like a TBB backend that we can use as default CPU backend instead of openmp sequential, which is basically a hack.

Yep, no worries, the comment above `omp_sequential_invocation` makes this clear.

Reply:
>  So I can see that if OpenMP is selected as one of the targets, it uses the full -fopenmp rather than getting the link flag from omp_sequential_invocation. But even then it does not use the OpenMP of the user-requested host compiler, at least in my testing. fopenmp always presumes libomp unless I modify it to -fopenmp=libgomp, which the HIP backend then accepts. Thus the idea to pin the OpenMP flags to match the host compiler.

Okay - I can see the use case of maybe wanting to use clang with `libgomp`. But this is a truly new feature as `syclcc` has always just used `-fopenmp` so I don't think anything has changed here with this PR.
As far as I'm aware @sbalint98 has started some work to expose the compiler cxx flags and link lines for all backends (similarly to e.g. `CUDA_LINK_LINE` that we have already). Then the user could change this to `-fopenmp=libgomp` if gomp should be used. Do you think that there needs to be something in addition to that?

I'm not sure if using the OpenMP library of the host compiler is the right thing to do. The user might have different requirements when only targeting CPU (which would then invoke the host compiler) vs when also compiling for GPU. For example, a user might want to use Intel C++ and Intel OpenMP for the best CPU performance when targeting CPUs exclusively, but when also compiling for GPUs, good compatibility with clang might be more important.

> Cmake is finding my GCC OpenMP so it's not failing, but not identifying the lack of LLVM OpenMP.

hm.. yeah maybe we should require LLVM OpenMP specifically in cmake when a GPU backend is enabled.

Reply:
> > So I can see that if OpenMP is selected as one of the targets, it uses the full -fopenmp rather than getting the link flag from omp_sequential_invocation. But even then it does not use the OpenMP of the user-requested host compiler, at least in my testing. fopenmp always presumes libomp unless I modify it to -fopenmp=libgomp, which the HIP backend then accepts. Thus the idea to pin the OpenMP flags to match the host compiler.
> 
> Okay - I can see the use case of maybe wanting to use clang with `libgomp`. But this is a truly new feature as `syclcc` has always just used `-fopenmp` so I don't think anything has changed here with this PR.
> As far as I'm aware @sbalint98 has started some work to expose the compiler cxx flags and link lines for all backends (similarly to e.g. `CUDA_LINK_LINE` that we have already). Then the user could change this to `-fopenmp=libgomp` if gomp should be used. Do you think that there needs to be something in addition to that?
> 
> I'm not sure if just using the OpenMP library of the host compiler is the right thing to do. The user might have different requirements when only targeting CPU (which would then invoke the host compiler) vs when also compiling for GPU. For example, a user might want to use Intel C++ and Intel OpenMP for the best CPU performance when targeting CPUs exclusively, but when also compiling for GPUs, good compatibility with clang might be more important.

Before merging this PR, I was able to build for the OpenMP and ROCm backend with no modification of flags without having LLVM OpenMP. I haven't done any diagnosing to figure out what has changed in terms of the underlying clang invocations though.

Exposing the compile and link flags makes good sense I think. More flexibility lets the end user do deployment-specific workarounds in weird cases like mine. Either way, compatibility with at least clang, gcc, and icc would definitely help the flexibility of our deployment. (We even have some optional components built for PGI that fallback on OpenMP if OpenACC isn't available, so ensuring consistency of implementation with the syclcc-compiled objects is a high priority.)

Reply:
> Before merging this PR, I was able to build for the OpenMP and ROCm backend with no modification of flags without having LLVM OpenMP. I haven't done any diagnosing to figure out what has changed in terms of the underlying clang invocations though.

What this PR changes is that kernels are now always also compiled for CPU, even if only targeting GPU explicitly (in order to meet the specification's requirement to have the host device always available and working). Previously, CPU was only available if (and only if) you explicitly target CPU.

You have presumably been using `g++` as CPU compiler, in which case `-fopenmp` just uses libgomp. But if you try using clang as host compiler, you will get the same behavior as we are seeing now because it also just compiles with `-fopenmp`. So, compiling for CPU with clang but using libgomp is something that has never worked before. You might just not have noticed it because only now do we also need the CPU backend when compiling for GPUs, which always requires compilation with clang, so you cannot get around clang-based compilation for CPU anymore.